### PR TITLE
added dummy audience and features to frontend fixes #2843

### LINKED ISSUE #2843: Feature/Audience for Rapid Intake UI
Once https://github.com/mozilla/experimenter/issues/2842 is done we can hook those fields up to the new Rapid UI.  The fields are already built they just need to be hooked up.

### PR DESCRIPTION
**Pull request by tiftran**

### CODE REVIEW

File: app/experimenter/experiments/api/v3/serializers.py [hunk 1]
```diff
@@ -15,6 +15,12 @@ class ExperimentRapidSerializer(ChangelogSerializerMixin, serializers.ModelSeria
         default=Experiment.BUGZILLA_RAPID_EXPERIMENT_TEMPLATE
     )
     objectives = serializers.CharField(required=True)
+    features = serializers.ListField(
+        required=True, child=serializers.CharField(), allow_empty=False
```
Comment (jaredlockhart) [inline app/experimenter/experiments/api/v3/serializers.py:19]: Is it possible to pass the list of FEATURE_CHOICES into this charfield?

File: app/experimenter/experiments/constants.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def FEATURE_TYPE_CHOICES(cls):  # pragma: no cover
 
     RAPID_FEATURE_CHOICES = (("FEATURE 1", "FEATURE 1"), ("FEATURE 2", "FEATURE 2"))
 
-    RAPID_AUDIENCE_CHOICES = (("AUDIENCE 1", "AUDIENCE1"), ("AUDIENCE 2", "AUDIENCE 2"))
+    RAPID_AUDIENCE_CHOICES = (("AUDIENCE 1", "AUDIENCE 1"), ("AUDIENCE 2", "AUDIENCE 2"))
```
Comment (jaredlockhart) [inline app/experimenter/experiments/constants.py:70]: ;)

File: app/experimenter/experiments/tests/api/v3/test_serializers.py [hunk 3]
```diff
@@ -122,13 +122,17 @@ def test_serializer_creates_changelog_for_updates(self):
...
+            features=["FEATURE 1"],
             public_description=Experiment.BUGZILLA_RAPID_EXPERIMENT_TEMPLATE,
         )
 
         self.assertEqual(experiment.changes.count(), 1)
         data = {
             "name": "changing the name",
             "objectives": "changing objectives",
+            "audience": "AUDIENCE 1",
+            "features": ["FEATURE 1"],
```
Comment (jaredlockhart) [inline app/experimenter/experiments/tests/api/v3/test_serializers.py:135]: I'd like to see some failure tests for like

- invalid audience choice
- missing audience choice
- invalid feature choices
- empty feature choices

File: app/experimenter/static/rapid/components/experiments/ExperimentDetails.tsx [hunk 4]
```diff
@@ -37,8 +37,8 @@ const ExperimentDetails: React.FC = () => {
         </div>
       </LabelledRow>
       <LabelledRow label="Hypothesis" value={data.objectives} />
-      <LabelledRow label="Feature" />
-      <LabelledRow label="Audience" />
+      <LabelledRow label="Feature" value={data.features.toString()} />
+      <LabelledRow label="Audience" value={data.audience} />
```
Comment (jaredlockhart) [inline app/experimenter/static/rapid/components/experiments/ExperimentDetails.tsx:41]: Are these the labels or the values? We should be showing the labels, not the values.

File: app/experimenter/static/rapid/components/forms/ExperimentForm.tsx [hunk 5]
```diff
@@ -37,6 +37,37 @@ const ExperimentForm: React.FC = () => {
...
+
+  const handleAudienceChange = (audienceValue) => {
+    dispatch({
+      type: ExperimentReducerActionType.UPDATE_STATE,
+      state: {
+        ...formData,
+        audience: audienceValue.value,
+      },
+    });
+  };
```
Comment (jaredlockhart) [inline app/experimenter/static/rapid/components/forms/ExperimentForm.tsx:69]: You can't reuse the `handleChange` below for this?

File: app/experimenter/experiments/tests/api/v3/test_serializers.py (hunk 3 above)
Comment (tiftran) [reply app/experimenter/experiments/tests/api/v3/test_serializers.py:135]: will do

File: app/experimenter/static/rapid/components/experiments/ExperimentDetails.tsx (hunk 4 above)
Comment (tiftran) [reply app/experimenter/static/rapid/components/experiments/ExperimentDetails.tsx:41]: oooh you're right you're right

File: app/experimenter/static/rapid/components/forms/ExperimentForm.tsx (hunk 5 above)
Comment (tiftran) [reply app/experimenter/static/rapid/components/forms/ExperimentForm.tsx:69]: we could if we decide to save feature and audience values as 
`{"value": "feature value", "label": "feature label" }`?
Do we prefer that?

Comment (jaredlockhart) [reply app/experimenter/static/rapid/components/forms/ExperimentForm.tsx:69]: Hrmmm yeah we have to store the whole object somewhere to display the right values, but we only want to actually store the value in the data object that gets posted back to the server.  Maybe there's some way we can hide that logic inside the implementation of the Select component?

**Comment by jaredlockhart** [comment]
Screenshot 2020-07-02 17 21 09

This isn't the worst but I wonder if there's a way to make the serializer say something more help than that

Maybe if we sent null instead of `""` as the default?  Not sure if this is a frontend or backend thing

**Comment by jaredlockhart** [comment]
Oh also `Description of the audience` doesn't feel like the right msg since to me that suggests entering text rather than choosing a selection, but that's what's in the figma so we should change it there.  We really need a way to file bugs against the figma.

**Comment by jaredlockhart** [comment]
Oh I thought we were gonna take `trigger` out in this PR, but we can also do that as a followup.

**Comment by jaredlockhart** [comment]
Screenshot 2020-07-02 17 34 39

^ Spacing between feature labels

Maybe we can even put them on separate lines?

File: app/experimenter/static/rapid/components/forms/ExperimentForm.tsx [hunk 6]
```diff
@@ -37,6 +37,37 @@ const ExperimentForm: React.FC = () => {
...
+    { value: "FEATURE 1", label: "FEATURE 1" },
+    { value: "FEATURE 2", label: "FEATURE 2" },
+  ];
+
+  const audienceOptions = [
+    { value: "AUDIENCE 1", label: "AUDIENCE 1" },
+    { value: "AUDIENCE 2", label: "AUDIENCE 2" },
+  ];
+
+  const handlefeatureChange = (featureValues) => {
```
Comment (pdehaan) [inline app/experimenter/static/rapid/components/forms/ExperimentForm.tsx:50]: nit: should this be `handleFeatureChange()` (upper "F") to be consistent with `handleAudienceChange()` below?

File: app/experimenter/static/rapid/components/forms/ExperimentForm.tsx [hunk 7]
```diff
@@ -117,10 +148,19 @@ const ExperimentForm: React.FC = () => {
...
-          options={[]}
+          options={featureOptions}
+          value={
+            formData.features
+              ? featureOptions.filter((option) =>
+                  formData.features.includes(option.value),
+                )
+              : null
+          }
+          onOptionChange={handlefeatureChange}
```
Comment (pdehaan) [inline app/experimenter/static/rapid/components/forms/ExperimentForm.tsx:163]: and `handleFeatureChange` here if we change the function name above.

**Comment by tiftran** [comment]
new tests written....

Screen Shot 2020-07-02 at 8 26 26 PM
 not sure if this looks better....

File: app/experimenter/static/rapid/components/forms/ExperimentFormOptions.tsx [hunk 8]
```diff
@@ -0,0 +1,9 @@
+export const audienceOptions = [
+  { label: "AUDIENCE 1", value: "AUDIENCE 1" },
+  { label: "AUDIENCE 2", value: "AUDIENCE 2" },
+];
+
+export const featureOptions = [
+  { label: "FEATURE 1", value: "FEATURE 1" },
+  { label: "FEATURE 2", value: "FEATURE 2" },
+];
```
Comment (pdehaan) [inline app/experimenter/static/rapid/components/forms/ExperimentFormOptions.tsx:9]: Is the plan that these would be hardcoded in Experimenter, or eventually be pulled from the nimbus-shared module?

Oh, nevermind. Now I remember seeing some recent PRs in https://github.com/mozilla/nimbus-shared/pull/54 and https://github.com/mozilla/nimbus-shared/pull/55 for adding Audiences and Features.

Comment (tiftran) [reply app/experimenter/static/rapid/components/forms/ExperimentFormOptions.tsx:9]: yup yup, just temporary

File: app/experimenter/static/rapid/components/experiments/ExperimentDetails.tsx [hunk 9]
```diff
@@ -14,13 +19,23 @@ const LabelledRow: React.FC<{ label: string; value?: string }> = ({
         {label}
       </label>
       <span className="col-10">
-        <input readOnly className="w-100" type="text" value={value || ""} />
+        <input
+          readOnly
+          className="w-100"
+          hidden={value ? false : true}
```
[the remark this answers is not in the document]
Comment (tiftran) [reply app/experimenter/static/rapid/components/experiments/ExperimentDetails.tsx:25]: I think i'll be reverting it the way it was before, given that @jaredlockhart  it looks confusing

File: app/experimenter/static/rapid/components/forms/XSelect.tsx [hunk 10]
```diff
@@ -1,12 +1,12 @@
 import React from "react";
 import Select, {
   components,
+  // ActionMeta,
```
Comment (jaredlockhart) [inline app/experimenter/static/rapid/components/forms/XSelect.tsx:4]: What's this?

**Pull request merged**